pkg/model/firewall.go

```diff
@@ -42,72 +42,74 @@ type FirewallModelBuilder struct {
-	{
-		t := &awstasks.SecurityGroup{
-			Name:             s(b.SecurityGroupName(kops.InstanceGroupRoleNode)),
-			Lifecycle:        b.Lifecycle,
-			VPC:              b.LinkToVPC(),
-			Description:      s("Security group for nodes"),
-			RemoveExtraRules: []string{"port=22"},
-		}
-		c.AddTask(t)
-	}
+func (b *FirewallModelBuilder) buildNodeRules(c *fi.ModelBuilderContext) (map[string]*awstasks.SecurityGroup, error) {
 
-	// Allow full egress
-	{
-		t := &awstasks.SecurityGroupRule{
-			Name:          s("node-egress"),
-			Lifecycle:     b.Lifecycle,
-			SecurityGroup: b.LinkToSecurityGroup(kops.InstanceGroupRoleNode),
-			Egress:        fi.Bool(true),
-			CIDR:          s("0.0.0.0/0"),
-		}
-		c.AddTask(t)
+	nodeGroups, err := CreateSecurityGroups(b.InstanceGroups, kops.InstanceGroupRoleNode, b.ClusterName(), b.Lifecycle, b.LinkToVPC(), c)
+	if err != nil {
+		return nil, err
 	}
 
-	// Nodes can talk to nodes
-	{
-		t := &awstasks.SecurityGroupRule{
-			Name:          s("all-node-to-node"),
-			Lifecycle:     b.Lifecycle,
-			SecurityGroup: b.LinkToSecurityGroup(kops.InstanceGroupRoleNode),
-			SourceGroup:   b.LinkToSecurityGroup(kops.InstanceGroupRoleNode),
+	for nodeGroupName, secGroup := range nodeGroups {
+		suffix := GetGroupSuffix(nodeGroupName, nodeGroups)
+		// Allow full egress
+		{
+			t := &awstasks.SecurityGroupRule{
+				Name:          s(fmt.Sprintf("node-egress%s", suffix)),
+				Lifecycle:     b.Lifecycle,
+				SecurityGroup: secGroup,
+				Egress:        fi.Bool(true),
+				CIDR:          s("0.0.0.0/0"),
+			}
+			c.AddTask(t)
 		}
-		c.AddTask(t)
-	}
 
-	// Pods running in Nodes could need to reach pods in master/s
-	if b.Cluster.Spec.Networking != nil && b.Cluster.Spec.Networking.AmazonVPC != nil {
-		// Nodes can talk to masters
+		// Nodes can talk to nodes
 		{
 			t := &awstasks.SecurityGroupRule{
-				Name:          s("all-nodes-to-master"),
+				Name:          s(fmt.Sprintf("all-node-to-node%s", suffix)),
 				Lifecycle:     b.Lifecycle,
-				SecurityGroup: b.LinkToSecurityGroup(kops.InstanceGroupRoleMaster),
-				SourceGroup:   b.LinkToSecurityGroup(kops.InstanceGroupRoleNode),
+				SecurityGroup: secGroup,
+				SourceGroup:   secGroup,
 			}
 			c.AddTask(t)
 		}
-	}
 
-	// We _should_ block per port... but:
-	// * It causes e2e tests to break
-	// * Users expect to be able to reach pods
-	// * If users are running an overlay, we punch a hole in it anyway
-	//b.applyNodeToMasterAllowSpecificPorts(c)
-	b.applyNodeToMasterBlockSpecificPorts(c)
+		// Pods running in Nodes could need to reach pods in master/s
+		if b.Cluster.Spec.Networking != nil && b.Cluster.Spec.Networking.AmazonVPC != nil {
+			// Nodes can talk to masters
+			{
+				t := &awstasks.SecurityGroupRule{
+					Name:          s(fmt.Sprintf("all-nodes-to-master%s", suffix)),
+					Lifecycle:     b.Lifecycle,
+					SecurityGroup: b.LinkToSecurityGroup(kops.InstanceGroupRoleMaster),
```

Comment:
This should be the same logic, right?

One idea might be to have a function which takes a template SecurityGroupRule, and expands it with all the SGs - may be simpler